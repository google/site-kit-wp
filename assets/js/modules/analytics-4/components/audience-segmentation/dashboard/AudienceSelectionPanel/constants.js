/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
