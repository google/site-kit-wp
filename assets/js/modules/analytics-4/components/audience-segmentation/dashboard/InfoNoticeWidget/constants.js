/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
