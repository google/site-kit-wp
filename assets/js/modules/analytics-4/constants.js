/**
 * GA4 module constants.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const ACTIVATION_STEP_REMINDER = 0;
export const ACTIVATION_STEP_SETUP = 1;
export const ACTIVATION_STEP_SUCCESS = 2;
export const ACTIVATION_STEP_IN_PROGRESS = 3;

export const ACTIVATION_ACKNOWLEDGEMENT_TOOLTIP_STATE_KEY =
	'activation-acknowledgement-tooltip-state';

export const GA4_ACTIVATION_BANNER_STATE_KEY = 'ga4-activation-banner-state';

export const ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY =
	'enhanced-measurement-activation-banner-tooltip-state';
export const LEGACY_ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY =
	'enhanced-measurement-activation-banner-dismissed-item';

export const REPORT_ARGS_SELECTED_METRIC_KEY = '_r.explorerCard..selmet';
export const REPORT_ARGS_SELECTED_DIMENSION_KEY = '_r.explorerCard..seldim';
export const REPORT_ARGS_DATA_FILTERS_KEY = '_r..dataFilters';
export const REPORT_ARGS_NAV_KEY = '_r..nav';

export const KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY =
	'key-metrics-connect-ga4-cta-widget';

/**
 * Date that Site Kit will automatically switch to using GA4: September 25, 2023.
 *
 * With Site Kit's maximum date range being three months, once this period has passed since the UA cutoff date,
 * it will no longer be able to show any historical data for UA and will automatically switch to GA4.
 * For operational reasons, we automatically make the switch a few days before the three month cutoff point.
 *
 * @since 1.107.0
 * @see UA_CUTOFF_DATE.
 */
export const GA4_AUTO_SWITCH_DATE = '2023-09-25';
export const MODULE_SLUG_ANALYTICS_4 = 'analytics-4';
