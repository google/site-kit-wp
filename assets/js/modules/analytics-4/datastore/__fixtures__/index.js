/**
 * GA4 Datastore Fixtures.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as accountSummaries } from './account-summaries.json';
export { default as audiences } from './audiences.json';
export { default as availableAudiences } from './available-audiences.json';
export { default as container } from './container.json';
export { default as containerE2E } from './container-e2e.json';
export { default as containerDestinations } from './container-destinations.json';
export { default as conversionEvents } from './conversion-events.json';
export { default as createProperty } from './create-property.json';
export { default as createWebDataStream } from './create-webdatastream.json';
export { default as defaultSettings } from './default-settings.json';
export { default as defaultEnhancedMeasurementSettings } from './default-enhanced-measurement-settings.json';
export { default as pageTitles } from './page-titles.json';
export { default as properties } from './properties.json';
export { default as report } from './report.json';
export { default as pivotReport } from './pivot-report.json';
export { default as webDataStreams } from './webdatastreams.json';
export { default as webDataStreamsBatch } from './webdatastreams-batch.json';
export { default as webDataStreamsBatchSetup } from './webdatastreams-batch-setup.json';
export { default as googleTagSettings } from './google-tag-settings.json';
