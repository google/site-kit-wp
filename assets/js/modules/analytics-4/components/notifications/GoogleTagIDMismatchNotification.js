/**
 * GoogleTagIDMismatchNotification component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';
import { useCallback } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '../../constants';
import { isValidMeasurementID } from '../../utils/validation';
import { CORE_MODULES } from '../../../../googlesitekit/modules/datastore/constants';
import { getBestTagID } from '../../utils/google-tag';
import BannerNotification, {
	TYPES,
} from '@/js/googlesitekit/notifications/components/layout/BannerNotification';
import { ProgressBar } from 'googlesitekit-components';

export default function GoogleTagIDMismatchNotification( {
	id,
	Notification,
} ) {
	const currentAnalyticsConfig = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getSettings()
	);

	const googleTagContainerDestinations = useSelect(
		( select ) =>
			currentAnalyticsConfig?.googleTagAccountID &&
			currentAnalyticsConfig?.googleTagContainerID &&
			select( MODULES_ANALYTICS_4 ).getGoogleTagContainerDestinations(
				currentAnalyticsConfig.googleTagAccountID,
				currentAnalyticsConfig.googleTagContainerID
			)
	);

	const currentAnalyticsProperty = useSelect(
		( select ) =>
			currentAnalyticsConfig?.propertyID &&
			select( MODULES_ANALYTICS_4 ).getProperty(
				currentAnalyticsConfig.propertyID
			)
	);

	const newAnalyticsConfig = useSelect( ( select ) => {
		if ( ! Array.isArray( googleTagContainerDestinations ) ) {
			return null;
		}

		const destinationIDs = googleTagContainerDestinations.map(
			// eslint-disable-next-line sitekit/acronym-case
			( destination ) => destination.destinationId
		);

		const validGA4MeasurementIDs =
			destinationIDs.filter( isValidMeasurementID );

		if ( validGA4MeasurementIDs.length === 0 ) {
			return null;
		}

		return select( MODULES_ANALYTICS_4 ).getAnalyticsConfigByMeasurementIDs(
			validGA4MeasurementIDs
		);
	} );

	const newAnalyticsProperty = useSelect(
		( select ) =>
			newAnalyticsConfig &&
			newAnalyticsConfig.propertyID &&
			select( MODULES_ANALYTICS_4 ).getProperty(
				newAnalyticsConfig?.propertyID
			)
	);

	const newGoogleTagID = useSelect( ( select ) => {
		const tagIDs =
			currentAnalyticsConfig?.measurementID &&
			select( MODULES_ANALYTICS_4 ).getGoogleTagContainer(
				currentAnalyticsConfig?.measurementID
			)?.tagIds; // eslint-disable-line sitekit/acronym-case

		if ( Array.isArray( tagIDs ) ) {
			return getBestTagID( tagIDs, currentAnalyticsConfig.measurementID );
		}
	} );

	const isDoingSubmitChanges = useSelect( ( select ) =>
		select( CORE_MODULES ).isDoingSubmitChanges( MODULE_SLUG_ANALYTICS_4 )
	);

	const {
		setPropertyID,
		setWebDataStreamID,
		setMeasurementID,
		updateSettingsForMeasurementID,
		submitChanges,
		setHasMismatchedGoogleTagID,
	} = useDispatch( MODULES_ANALYTICS_4 );

	const updateToNewAnalyticsConfig = useCallback( async () => {
		setPropertyID( newAnalyticsConfig?.propertyID );
		setWebDataStreamID( newAnalyticsConfig?.webDataStreamID );
		setMeasurementID( newAnalyticsConfig?.measurementID );
		await submitChanges();
		setHasMismatchedGoogleTagID( false );
	}, [
		setPropertyID,
		setWebDataStreamID,
		setMeasurementID,
		newAnalyticsConfig?.propertyID,
		newAnalyticsConfig?.webDataStreamID,
		newAnalyticsConfig?.measurementID,
		submitChanges,
		setHasMismatchedGoogleTagID,
	] );

	const updateGoogleTagConfig = useCallback( async () => {
		await updateSettingsForMeasurementID(
			currentAnalyticsConfig?.measurementID
		);
		await submitChanges();
		setHasMismatchedGoogleTagID( false );
	}, [
		updateSettingsForMeasurementID,
		currentAnalyticsConfig?.measurementID,
		submitChanges,
		setHasMismatchedGoogleTagID,
	] );

	if (
		googleTagContainerDestinations === undefined ||
		newAnalyticsProperty === undefined ||
		newGoogleTagID === undefined ||
		currentAnalyticsProperty === undefined
	) {
		return (
			<ProgressBar
				className="googlesitekit-banner__progress-bar"
				height={ 4 } // The progress bar height is set to 7px via CSS. This value removes margins within the ProgressBar component.
				indeterminate
			/>
		);
	}

	// If the current and new properties are the same, don't show the notification.
	// This can happen momentarily when the banner is actioned and the new property
	// has been saved, but the notification hasn't been dismissed / removed from the
	// queue yet. There is also a rare issue reported where hasMismatchedGoogleTagID
	// is set to true, but the new property is the same as the current one. This will
	// result in a perpetual loading banner on the screen if we returned the loading
	// bar. So it is safer to return null, as eventually, the syncGoogleTag action
	// will be called within an hour and reset this correctly.
	if (
		currentAnalyticsProperty?._id &&
		newAnalyticsProperty?._id &&
		currentAnalyticsProperty._id === newAnalyticsProperty._id
	) {
		// return null;
	}

	if ( newAnalyticsProperty ) {
		return (
			<Notification className="googlesitekit-publisher-win">
				<BannerNotification
					notificationID={ id }
					type={ TYPES.WARNING }
					title={ __(
						"Update Site Kit's Analytics configuration to continue seeing Analytics data on the dashboard",
						'google-site-kit'
					) }
					description={ sprintf(
						/* translators: 1: Current GA4 property name. 2: Current GA4 property ID. 3: Newly linked GA4 property name. 4: Newly linked GA4 property ID. */
						__(
							'The Google Tag on your site is no longer associated with your current Google Analytics property "%1$s (%2$s)". It is now recording metrics to another Google Analytics property "%3$s (%4$s)". If you want to continue seeing Analytics data in the Site Kit dashboard, we suggest you update Site Kitâ€™s Google Analytics configuration to show data for the property used in your Google Tag.',
							'google-site-kit'
						),
						currentAnalyticsProperty.displayName,
						currentAnalyticsProperty._id,
						newAnalyticsProperty.displayName,
						newAnalyticsProperty._id
					) }
					ctaButton={ {
						label: __( 'Use new property', 'google-site-kit' ),
						onClick: updateToNewAnalyticsConfig,
						inProgress: isDoingSubmitChanges,
					} }
					dismissButton={ {
						label: __(
							'Keep existing property',
							'google-site-kit'
						),
						onClick: updateGoogleTagConfig,
					} }
					dismissOptions={ {
						expiresInSeconds: 1,
					} }
					dismissOnCTAClick
				/>
			</Notification>
		);
	}

	if ( newGoogleTagID ) {
		return (
			<Notification className="googlesitekit-publisher-win">
				<BannerNotification
					notificationID={ id }
					type={ TYPES.WARNING }
					title={ __(
						'Your Google tag configuration has changed',
						'google-site-kit'
					) }
					description={ sprintf(
						/* translators: 1: Currently set Google Tag ID. 2: Newly linked Google Tag ID. 3: Current GA4 property name. 4: Current GA4 property ID. */
						__(
							'The Google tag for your Google Analytics configuration has changed from %1$s to %2$s. To keep using your current Google Analytics property "%3$s (%4$s)", you need to configure Site Kit to place the new Google tag %2$s instead.',
							'google-site-kit'
						),
						currentAnalyticsConfig.googleTagID,
						newGoogleTagID,
						currentAnalyticsProperty.displayName,
						currentAnalyticsProperty._id
					) }
					ctaButton={ {
						label: sprintf(
							/* translators: %s: Newly linked Google Tag ID. */
							__( 'Update Google tag to %s', 'google-site-kit' ),
							newGoogleTagID
						),
						onClick: updateGoogleTagConfig,
					} }
					dismissOptions={ {
						expiresInSeconds: 1,
					} }
					dismissOnCTAClick
				/>
			</Notification>
		);
	}
}
