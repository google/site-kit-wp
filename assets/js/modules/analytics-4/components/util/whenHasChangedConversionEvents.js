/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';

/**
 * Higher-Order Component to render wrapped components when changed conversion events are detected.
 *
 * @since n.e.x.t
 *
 * @return {Function} Enhancing function.
 */
export default function whenHasChangedConversionEvents() {
	return ( WrappedComponent ) => {
		function WhenHasChangedConversionEvents( props ) {
			const hasChangedConversionEvents = useSelect( ( select ) => {
				const {
					hasNewConversionReportingEvents,
					hasLostConversionReportingEvents,
				} = select( MODULES_ANALYTICS_4 );

				return (
					hasNewConversionReportingEvents() ||
					hasLostConversionReportingEvents()
				);
			}, [] );

			if ( hasChangedConversionEvents ) {
				return <WrappedComponent { ...props } />;
			}

			const { WidgetNull = null } = props;
			return WidgetNull && <WidgetNull />;
		}

		return WhenHasChangedConversionEvents;
	};
}
