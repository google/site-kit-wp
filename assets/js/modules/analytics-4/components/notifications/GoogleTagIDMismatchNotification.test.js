/**
 * GoogleTagIDMismatchNotification component tests.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
	render,
	createTestRegistry,
	provideUserInfo,
	provideModules,
	provideUserAuthentication,
} from '../../../../../../tests/js/test-utils';
import GoogleTagIDMismatchNotification from './GoogleTagIDMismatchNotification';
import {
	GTM_SCOPE,
	MODULES_ANALYTICS_4,
} from '@/js/modules/analytics-4/datastore/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '@/js/modules/analytics-4/constants';
import { ANALYTICS_4_NOTIFICATIONS } from '@/js/modules/analytics-4';
import { withNotificationComponentProps } from '@/js/googlesitekit/notifications/util/component-props';
import * as fixtures from '@/js/modules/analytics-4/datastore/__fixtures__';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '@/js/googlesitekit/constants';

describe( 'GoogleTagIDMismatchNotification', () => {
	let registry;
	const GoogleTagIDMismatchNotificationComponent =
		withNotificationComponentProps( 'google-tag-id-mismatch' )(
			GoogleTagIDMismatchNotification
		);

	const notification = ANALYTICS_4_NOTIFICATIONS[ 'google-tag-id-mismatch' ];

	beforeEach( () => {
		registry = createTestRegistry();
		provideUserInfo( registry );
		provideUserAuthentication( registry, {
			grantedScopes: [ GTM_SCOPE ],
		} );
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: MODULE_SLUG_ANALYTICS_4,
			},
		] );

		const currentMeasurementID = 'G-2B7M8YQ1K6';

		const currentAnalyticsSettingsMock = {
			ownerID: 1,
			propertyID: '1000',
			webDataStreamID: '2000',
			measurementID: currentMeasurementID,
			useSnippet: true,
			googleTagID: 'GT-123',
			googleTagLastSyncedAtMs: 0,
		};
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setSettings( currentAnalyticsSettingsMock );
		const property = fixtures.properties[ 0 ];
		const propertyID = property._id;
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( property, { propertyID } );

		const containerMock = fixtures.container[ currentMeasurementID ];
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetGoogleTagContainer( containerMock, {
				measurementID: currentMeasurementID,
			} );

		const newAnalyticsPropertyID = '1122334455';
		const accountSummaries = [
			{
				account: 'accounts/123456',
				name: 'accountSummaries/987654321',
				propertySummaries: [
					{
						property: `properties/${ newAnalyticsPropertyID }`,
						_id: newAnalyticsPropertyID,
					},
				],
				_id: '123456',
			},
		];

		const propertyIDs = accountSummaries
			.map( ( { propertySummaries } ) =>
				propertySummaries.map( ( { _id } ) => _id )
			)
			.reduce( ( acc, propIDs ) => [ ...acc, ...propIDs ], [] );

		const datastreams = {
			[ newAnalyticsPropertyID ]: [
				{
					_id: '110',
					webStreamData: {
						defaultUri: 'http://example-1.test', // eslint-disable-line sitekit/acronym-case
						measurementId: currentMeasurementID, // eslint-disable-line sitekit/acronym-case
					},
				},
			],
		};
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries( {
			accountSummaries,
			nextPageToken: null,
		} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( datastreams, {
				propertyIDs,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.finishResolution( 'getWebDataStreamsBatch', [ propertyIDs ] );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( fixtures.properties[ 1 ], {
				propertyID: newAnalyticsPropertyID,
			} );
	} );

	it( 'should render the "alternative GA4 Config found" notification variation', async () => {
		const gtmAccountID = '6065484567';
		const gtmContainerID = '98369876';
		const containerDestinationsMock =
			fixtures.containerDestinations[ gtmAccountID ][ gtmContainerID ];

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setGoogleTagAccountID( gtmAccountID );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setGoogleTagContainerID( gtmContainerID );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetGoogleTagContainerDestinations(
				containerDestinationsMock,
				{
					gtmAccountID,
					gtmContainerID,
				}
			);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveHasMismatchGoogleTagID( true );

		const { container, waitForRegistry } = render(
			<GoogleTagIDMismatchNotificationComponent />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the "alternative GA4 Config not found" notification variation', async () => {
		const gtmAccountID = '6065484567';
		const gtmContainerID = '98369876';

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setGoogleTagAccountID( gtmAccountID );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setGoogleTagContainerID( gtmContainerID );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetGoogleTagContainerDestinations( [], {
				gtmAccountID,
				gtmContainerID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveHasMismatchGoogleTagID( true );

		const { container, waitForRegistry } = render(
			<GoogleTagIDMismatchNotificationComponent />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );

	describe( 'checkRequirements', () => {
		it( 'is active when all conditions are met', async () => {
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveHasMismatchGoogleTagID( true );

			const isActive = await notification.checkRequirements(
				registry,
				VIEW_CONTEXT_MAIN_DASHBOARD
			);
			expect( isActive ).toBe( true );
		} );

		it( 'is not active when analytics-4 is not connected', async () => {
			provideModules( registry, [
				{
					active: false,
					connected: false,
					slug: MODULE_SLUG_ANALYTICS_4,
				},
			] );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveHasMismatchGoogleTagID( true );

			const isActive = await notification.checkRequirements(
				registry,
				VIEW_CONTEXT_MAIN_DASHBOARD
			);
			expect( isActive ).toBe( false );
		} );

		it( 'is not active when current logged in user is not analytics-4 module owner', async () => {
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setSettings( { ownerID: 2 } );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveHasMismatchGoogleTagID( true );

			const isActive = await notification.checkRequirements(
				registry,
				VIEW_CONTEXT_MAIN_DASHBOARD
			);
			expect( isActive ).toBe( false );
		} );

		it( 'is not active when user does not have GTM scope', async () => {
			provideUserAuthentication( registry );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveHasMismatchGoogleTagID( true );

			const isActive = await notification.checkRequirements(
				registry,
				VIEW_CONTEXT_MAIN_DASHBOARD
			);
			expect( isActive ).toBe( false );
		} );

		it( 'is not active when mismatched Google Tag ID is false', async () => {
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveHasMismatchGoogleTagID( false );

			const isActive = await notification.checkRequirements(
				registry,
				VIEW_CONTEXT_MAIN_DASHBOARD
			);
			expect( isActive ).toBe( false );
		} );
	} );
} );
