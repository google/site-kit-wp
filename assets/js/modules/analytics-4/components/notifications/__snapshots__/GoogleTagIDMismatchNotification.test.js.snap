// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleTagIDMismatchNotification should render the "alternative GA4 Config found" notification variation 1`] = `
<div>
  <section
    class="googlesitekit-publisher-win"
    id="google-tag-id-mismatch"
  >
    <div
      class="googlesitekit-banner-notification googlesitekit-banner-notification--warning"
    >
      <div
        class="mdc-layout-grid googlesitekit-page-content"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-banner"
            >
              <div
                class="googlesitekit-banner__content"
              >
                <p
                  class="googlesitekit-banner__title"
                >
                  Update Site Kit's Analytics configuration to continue seeing Analytics data on the dashboard
                </p>
                <div
                  class="googlesitekit-banner__description"
                >
                  The Google Tag on your site is no longer associated with your current Google Analytics property "Test GA4 Property (1000)". It is now recording metrics to another Google Analytics property "Another GA4 Property (1001)". If you want to continue seeing Analytics data in the Site Kit dashboard, we suggest you update Site Kitâ€™s Google Analytics configuration to show data for the property used in your Google Tag.
                   
                </div>
                <div
                  class="googlesitekit-notice__action"
                >
                  <button
                    class="mdc-button googlesitekit-banner__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Use new property
                    </span>
                  </button>
                  <button
                    class="mdc-button mdc-button--tertiary"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Keep existing property
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="googlesitekit-banner__svg-wrapper googlesitekit-banner__svg-wrapper--center"
                style="background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PC9zdmc+);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`GoogleTagIDMismatchNotification should render the "alternative GA4 Config not found" notification variation 1`] = `
<div>
  <section
    class="googlesitekit-publisher-win"
    id="google-tag-id-mismatch"
  >
    <div
      class="googlesitekit-banner-notification googlesitekit-banner-notification--warning"
    >
      <div
        class="mdc-layout-grid googlesitekit-page-content"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-banner"
            >
              <div
                class="googlesitekit-banner__content"
              >
                <p
                  class="googlesitekit-banner__title"
                >
                  Your Google tag configuration has changed
                </p>
                <div
                  class="googlesitekit-banner__description"
                >
                  The Google tag for your Google Analytics configuration has changed from GT-123 to GT-NBQN9V2. To keep using your current Google Analytics property "Test GA4 Property (1000)", you need to configure Site Kit to place the new Google tag GT-NBQN9V2 instead.
                   
                </div>
                <div
                  class="googlesitekit-notice__action"
                >
                  <button
                    class="mdc-button googlesitekit-banner__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Update Google tag to GT-NBQN9V2
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="googlesitekit-banner__svg-wrapper googlesitekit-banner__svg-wrapper--center"
                style="background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PC9zdmc+);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
